# SITypes: Scientific Units and Dimensional Analysis

## Introduction

Physical quantities require both a numerical value and a unit. For example, measuring time during a race, temperature in a patient's mouth, or mass of cheese all need units to be meaningful. Physical quantities can be scalars, vectors, matrices, or higher-dimensional objects, and with units, they acquire additional mathematical behaviors that numerical values alone lack.

Units enforce dimensional consistency: quantities can only be converted between units of the same dimensionality, added/subtracted only when dimensionally compatible, and transcendental functions require dimensionless arguments.

Physical quantities relate through physical laws, leading to Joseph Fourier's insight in 1822 that all quantities can be derived from a smaller set of fundamental reference quantities. While theoretically this could be reduced to just Planck length and speed of light, practically seven fundamental SI quantities are used: (1) length, (2) mass, (3) time, (4) electric current, (5) thermodynamic temperature, (6) amount of substance, and (7) luminous intensity.

## About SITypes

This notebook demonstrates SITypes through RMNpy, a Python library implementing these fundamental concepts. SITypes provides:

- Creating and manipulating scientific units
- Working with scalars (values with units)  
- Dimensional analysis and validation
- Unit conversions
- Complex calculations with automatic unit tracking

The examples below show how SITypes enforces mathematical rules governing physical quantities, preventing common errors while maintaining full expressiveness for complex physics and engineering problems.

## Setup and Imports

SITypes provides three fundamental classes for working with physical quantities:

1. **Dimensionality** - Represents the dimensional structure of physical quantities (e.g., Length, Mass, Time)
2. **Unit** - Represents specific units of measurement (e.g., meters, kilograms, seconds)  
3. **Scalar** - Represents values with units, combining numerical values with their units

We'll explore each of these in order, building up from fundamental concepts to practical applications.

## 1. Dimensionality

Dimensionality represents the fundamental structure of physical quantities in terms of the seven SI base dimensions. In the International System of Units (SI), seven reference quantities are used to define seven dimensions whose symbols are given in the table below:

In [74]:
# Import the SITypes dimensional analysis components
from rmnpy.sitypes import Dimensionality
from rmnpy.sitypes import quantity as q

# List of fundamental quantity constants using the cleaner aliases
fundamental_quantity_constants = [
    q.Length, 
    q.Mass, 
    q.Time, 
    q.Current, 
    q.Temperature,
    q.Amount, 
    q.LuminousIntensity
]

print("SI Fundamental Dimensions Table:")
print("=" * 60)
print(f"{'Reference Quantity':<30} | {'Dimensionality Symbol'}")
print("-" * 60)

for quantity_const in fundamental_quantity_constants:
    # Get the name directly from the constant
    name = str(quantity_const)
    dim = Dimensionality.for_quantity(quantity_const)
    print(f"{name:<30} | {dim}")

print(f"\n✓ RMNpy SITypes components imported successfully")
print(f"✓ Quantity module loaded with {len(q.get_all_quantity_names())} available quantities")
print(f"✓ Example quantities: {q.Length}, {q.Mass}, {q.Time}")

SI Fundamental Dimensions Table:
Reference Quantity             | Dimensionality Symbol
------------------------------------------------------------
length                         | L
mass                           | M
time                           | T
current                        | I
temperature                    | ϴ
amount                         | N
luminous intensity             | J

✓ RMNpy SITypes components imported successfully
✓ Quantity module loaded with 173 available quantities
✓ Example quantities: length, mass, time


### 1.1 Dimensionality Operations and Relationships

Dimensionalities can be combined through mathematical operations to create derived dimensions:

The dimensionality of any physical quantity, $q$, can then be expressed in terms of these seven reference dimensions in the form of a dimensional product:

$$\text{dim} \, q = \textsf{L}^\alpha \cdot \textsf{M}^\beta \cdot\textsf{T}^\gamma \cdot\textsf{I}^\delta \cdot\Theta^\epsilon\cdot \textsf{N}^\zeta \cdot\textsf{J}^\eta$$

where the lowercase Greek symbols represent integers called the dimensional exponents. The dimensionality of any physical quantity can be represented as a point in the space of dimensional exponents $(\alpha, \beta, \gamma, \delta, \epsilon, \zeta, \eta)$.

Physical quantities with different meanings can have the same dimensionality. For example, the thermodynamic quantities *entropy* and *heat capacity* are different physical quantities having the same physical dimensions. Only physical quantities with the same dimensionality can be added. With the operation of multiplication, the physical dimensions form a mathematical group.

There also exist numerous dimensionless quantities of importance, such as *plane angle* which has dimensionality of $\textsf{L}/\textsf{L}$, or *solid angle* with a dimensionality of $\textsf{L}^2/\textsf{L}^2$. It is sometimes necessary to distinguish between these different types of dimensionless quantities. Representing dimensionless dimensionalities requires us to split each dimension exponent into the numerator and denominator exponents. Thus, we redefine our set of dimensionalities to incorporate numerator and denominator exponents:

$$\text{dim} \, q = \left(\frac{\textsf{L}^{\alpha_+}}{\textsf{L}^{\alpha_-}} \right)
\cdot
\left(
  \frac{\textsf{M}^{\beta_+}}{\textsf{M}^{\beta_-}} \right)
\cdot
\left(
  \frac{\textsf{T}^{\gamma_+}}{\textsf{T}^{\gamma_-}} \right)
\cdot
\left(
  \frac{\textsf{I}^{\delta_+}}{\textsf{I}^{\delta_-}} \right)
\cdot
\left(
  \frac{\Theta^{\epsilon_+}}{\Theta^{\epsilon_-}} \right)
\cdot
\left(
  \frac{\textsf{N}^{\zeta_+}}{\textsf{N}^{\zeta_-}} \right)
\cdot
\left(\frac{\textsf{J}^{\eta_+}}{\textsf{J}^{\eta_-}} \right)$$

**Important distinction**: While there is an infinite set of possible derived dimensionalities through mathematical combinations, there is only a finite set of dimensionalities associated with recognized physical quantity names. SITypes provides 173 named quantities, but mathematical operations can create any dimensionality.

This mathematical framework forms the foundation of SITypes' dimensional analysis capabilities, as we'll demonstrate in the following examples.

In [75]:
# Demonstrate dimensionality operations and relationships
print("=== Dimensionality Operations ===")

# Create dimensionalities from quantity constants
length_dim = Dimensionality.for_quantity(q.Length)
time_dim = Dimensionality.for_quantity(q.Time) 
mass_dim = Dimensionality.for_quantity(q.Mass)

print(f"Basic dimensions:")
print(f"Length: {length_dim}")
print(f"Time: {time_dim}")
print(f"Mass: {mass_dim}")

# Combine dimensionalities using mathematical operations
velocity_dim = length_dim / time_dim
acceleration_dim = velocity_dim / time_dim
force_dim = mass_dim * acceleration_dim
energy_dim = force_dim * length_dim

print(f"\nDerived dimensions:")
print(f"Velocity (L/T): {velocity_dim}")
print(f"Acceleration (L/T²): {acceleration_dim}")
print(f"Force (M⋅L/T²): {force_dim}")
print(f"Energy (M⋅L²/T²): {energy_dim}")

# Create dimensionalities for other important quantities
area_dim = Dimensionality.for_quantity(q.Area)
volume_dim = Dimensionality.for_quantity(q.Volume)
frequency_dim = Dimensionality.for_quantity(q.Frequency)

print(f"\nOther important dimensions:")
print(f"Area (L²): {area_dim}")
print(f"Volume (L³): {volume_dim}")
print(f"Frequency (1/T): {frequency_dim}")

# Check dimensional consistency
print(f"\n=== Dimensional Consistency Checks ===")
print(f"Area = Length²: {area_dim == length_dim**2}")
print(f"Volume = Length³: {volume_dim == length_dim**3}")
print(f"Frequency = 1/Time: {frequency_dim == time_dim**-1}")
print(f"Energy = Force × Length: {energy_dim == force_dim * length_dim}")

print(f"\n✓ All {len(q.get_all_quantity_names())} quantity constants available through q.QuantityName")

=== Dimensionality Operations ===
Basic dimensions:
Length: L
Time: T
Mass: M

Derived dimensions:
Velocity (L/T): L/T
Acceleration (L/T²): L/T^2
Force (M⋅L/T²): M•L/T^2
Energy (M⋅L²/T²): M•L^2/T^2

Other important dimensions:
Area (L²): L^2
Volume (L³): L^3
Frequency (1/T): (1/T)

=== Dimensional Consistency Checks ===
Area = Length²: True
Volume = Length³: True
Frequency = 1/Time: True
Energy = Force × Length: True

✓ All 173 quantity constants available through q.QuantityName


## 2. Units

Inherent in measuring any physical quantity is a comparison to a previous measurement. What is most useful is the ratio of the new measurement to a previous measurement. For example, an ancient scientist might have used the cubit, an ancient unit of length, to measure the ratio of a large tree trunk's circumference to its diameter. Around 3000 BC in Egypt, a cubit was decreed to be the length of a forearm and hand. So, a scientist could make a string of 1 cubit in length using the distance from the back of their elbow to the tip of their middle finger and then use the string to measure the ratio:

$$\frac{l_\text{circumference}}{l_\text{diameter}} \approx 3.14$$

While another scientist with longer arms might have cut a longer string to be a cubit, the procedure for finding the ratio of a large tree trunk's circumference to its diameter will be the same, and the result is independent of the absolute length of the string used, i.e., independent of the units of length used.

Thus, we begin by representing a physical quantity, $q$, using the notation:

$$q = \{q \} \cdot [q]$$

where $\{q \}$ is the numerical value and $[q]$ is the reference unit symbol, i.e., a non-numerical string of characters, usually an abbreviation for the name of the unit.


### 2.1 SI Units

#### Coherent SI Base Units

The *coherent SI base (reference) units* form a set of seven units, given by the symbols:

$$[q]_\text{CBU} \in  [Q]_\text{CBU} = \left\{\text{m}, \text{kg}, \text{s}, \text{A}, \text{K}, \text{mol}, \text{cd} \right\}$$

#### SI Base Root Units

A minor complication is that the coherent base unit for mass in SI Units is defined as the kilogram, not the gram. For this reason, we define the set of seven base root units:

$$[q]_\text{BRU} \in  [Q]_\text{BRU} = \left\{\text{m}, \text{g}, \text{s}, \text{A}, \text{K}, \text{mol}, \text{cd} \right\}$$

#### SI Base Units

The set of *Coherent SI Base Units* only includes the seven SI base units mentioned above. The larger set of *SI Base Units* includes the Coherent SI Base Units, as well as all decimal multiples of the root units, created using the 20 SI prefixes with the root unit names and symbols. These prefixed and unprefixed symbols form the set, $[Q]_\text{BU}$, of 147 SI base units:

$$[q]_\text{BU} \in [Q]_\text{BU} = \left \{x_L\text{m}, x_M\text{g}, x_T\text{s}, x_I\text{A}, x_\Theta\text{K}, x_N\text{mol}, x_J\text{cd} \right \}$$

where $[Q]_\text{CBU}  \subseteq [Q]_\text{BU}$, $[Q]_\text{BRU} \subseteq [Q]_\text{BU}$, and $x_i$ indicates that the root unit symbol may be modified with one of the SI prefixes.

### 2.2 Creating and Manipulating Units

In [76]:
from rmnpy.sitypes import Unit

# Create basic units
meter = Unit("m")
second = Unit("s")
kilogram = Unit("kg")

print(f"Meter: {meter}")
print(f"Second: {second}")
print(f"Kilogram: {kilogram}")

# Combine units through mathematical operations
velocity_unit = meter / second
acceleration_unit = meter / (second * second)
force_unit = kilogram * meter / (second * second)

print(f"\nCombined units:")
print(f"Velocity unit: {velocity_unit}")
print(f"Acceleration unit: {acceleration_unit}")
print(f"Force unit: {force_unit}")

# Create complex units from expressions
energy_unit = Unit("kg*m^2/s^2")
pressure_unit = Unit("Pa")  # Pascal
frequency_unit = Unit("Hz")  # Hertz

print(f"\nComplex units:")
print(f"Energy unit: {energy_unit}")
print(f"Pressure unit: {pressure_unit}")
print(f"Frequency unit: {frequency_unit}")

# Show unit dimensionalities
print(f"\nUnit dimensionalities:")
print(f"Meter dimensionality: {meter.dimensionality}")
print(f"Force unit dimensionality: {force_unit.dimensionality}")
print(f"Energy unit dimensionality: {energy_unit.dimensionality}")

Meter: m
Second: s
Kilogram: kg

Combined units:
Velocity unit: m/s
Acceleration unit: m/s^2
Force unit: N

Complex units:
Energy unit: kg•m^2/s^2
Pressure unit: Pa
Frequency unit: Hz

Unit dimensionalities:
Meter dimensionality: L
Force unit dimensionality: M•L/T^2
Energy unit dimensionality: M•L^2/T^2


### 2.3 Special SI Unit Symbols

In the SI system, there is a set, $[Q]_\text{SU}$, of 22 coherent derived units contained within the $[Q]_\text{CDU}$ set that, for convenience, have their own special names and symbols. For example, the coherent derived SI unit: $\text{m}\cdot\text{kg}\cdot\text{s}^{-2}$, used for the derived quantity of force, is given the special name "newton" and replaced with the symbol "N".

The 22 coherent derived units with special SI units have the additional possibility of being modified by SI prefixes to create a set, $[Q]_{x\text{SU}}$, of 462 units. Some of the units in $[Q]_{x\text{SU}}$, however, may not have an equivalent in the set of derived SI units, $[Q]_\text{DU}$. That is, $[Q]_\text{DU} - [Q]_{x\text{SU}} \neq \emptyset$.

For example, the Sievert is equivalent to the coherent derived unit $\text{m}^2\cdot\text{s}^{-2}$, and:

$$[\text{Sv}] = [\text{m}^2\cdot\text{s}^{-2}] \subseteq  [Q]_\text{DU}$$

but the decisievert, $[\text{dSv}]$, which is one-tenth of the coherent derived unit $\text{m}^2\cdot\text{s}^{-2}$, has no equivalent unit in the set $[Q]_\text{DU}$. That is:

$$[\text{dSv}]  = 0.1 [\text{m}^2\cdot\text{s}^{-2}] \nsubseteq  [Q]_\text{DU}$$

On the other hand, the centiSievert, $[\text{cSv}]$, which is one-hundredth of the coherent derived unit $\text{m}^2\cdot\text{s}^{-2}$, has an equivalent unit in the set $[Q]_\text{DU}$:

$$[\text{cSv}]  = 0.01 [\text{m}^2\cdot\text{s}^{-2}] = [\text{dm}^2\cdot\text{s}^{-2}] = [\text{m}^2\cdot\text{das}^{-2}] \subseteq  [Q]_\text{DU}$$

With the introduction of special SI symbols, there is the additional possibility of derived quantities with derived units employing a special SI unit. Generally, however, no more than one special SI symbol appears in such a derived unit symbol, and the special SI symbol often appears as a linear term in the numerator. For example, the derived quantity dynamic viscosity can use the derived unit $[\text{Pa}\cdot\text{s}]$ as well as $[\text{m}^{-1}\cdot\text{kg}\cdot\text{s}^{-1}]$. The derived quantity of surface tension can use the derived unit $[\text{N}\cdot\text{m}^{-1}]$ as well as $[\text{kg}\cdot\text{s}^{-2}]$.

### 2.4 Non-SI Unit Symbols

Several units outside the International System of Units continue to be used in different science and engineering communities.   While it is straightforward to include non-SI unit symbols in the same manner as special SI unit symbols, one must avoid symbol and name collisions, particularly as some non-SI unit symbols employ SI prefixes.   

### 2.5 Physical Constant Symbols

It is also possible to use symbols for physical constants as unit symbols, although one has to be careful to avoid symbol collisions.  There are conflicts between unit symbols and commonly accepted symbols for physical constants, e.g., the planck constant and hour (h), or the newton gravitational constant and gauss (G).  Table~\ref{tb:PhysicalConstants} gives a list of suggested symbols for physical constants (giving priority to unit symbols) which avoid collision with accepted unit symbols.



### 2.6 Exploring All Possible Unit Token Symbols

SITypes provides access to all possible unit token symbols that can be used to construct units. This comprehensive library includes base units, prefixed units, special SI units, and derived units. The `get_unit_symbol_tokens_lib()` function returns the complete set of available symbols.

In [77]:
from rmnpy.sitypes import get_unit_symbol_tokens_lib

# Get all possible unit symbol tokens
all_symbols = get_unit_symbol_tokens_lib()
print(f"Total available unit symbol tokens: {len(all_symbols)}")

# Show the first 100 symbols to give an overview
print(f"\nFirst 100 unit symbol tokens:")
first_100 = list(all_symbols)[:100]
for i in range(0, len(first_100), 10):
    row = first_100[i:i+10]
    print("  " + "  ".join(f"{symbol:>6}" for symbol in row))

print(f"\n Use any of the {len(all_symbols)} symbols to create derived units with Unit('symbol')")

Total available unit symbol tokens: 1087

First 100 unit symbol tokens:
      ym      zm      am      fm      pm      nm      µm      mm      cm      dm
       m     dam      hm      km      Mm      Gm      Tm      Pm      Em      Zm
      Ym      yg      zg      ag      fg      pg      ng      µg      mg      cg
      dg       g     dag      hg      kg      Mg      Gg      Tg      Pg      Eg
      Zg      Yg      ys      zs      as      fs      ps      ns      µs      ms
      cs      ds       s     das      hs      ks      Ms      Gs      Ts      Ps
      Es      Zs      Ys      yA      zA      aA      fA      pA      nA      µA
      mA      cA      dA       A     daA      hA      kA      MA      GA      TA
      PA      EA      ZA      YA      yK      zK      aK      fK      pK      nK
      µK      mK      cK      dK       K     daK      hK      kK      MK      GK

 Use any of the 1087 symbols to create derived units with Unit('symbol')


In [78]:
# Create a complex Imperial energy unit (pound-force × area / time²)
energy_unit2 = Unit("lb*ft^2/min^2")
print(f"Imperial energy unit: {energy_unit2}")
print(f"Unit dimensionality: {energy_unit2.dimensionality}")

# Method 1: Convert to SI using string expression
si_scale_factor = energy_unit2.scale_to("g*m^2/s^2")
print(f"\nScale factor to SI (g⋅m²/s²): {si_scale_factor:.2e}")

# Method 2: Scale to coherent SI unit directly
coherent_si_scale = energy_unit2.scale_to_coherent_si
print(f"Scale to coherent SI: {coherent_si_scale:.2e}")


Imperial energy unit: ft^2•lb/min^2
Unit dimensionality: M•L^2/T^2

Scale factor to SI (g⋅m²/s²): 4.23e+13
Scale to coherent SI: 4.23e+10


## 3. Scalars

Scalars combine numerical values with units to represent physical quantities. They provide the complete picture: both the magnitude and the unit of measurement. Scalars are the primary way to work with physical quantities in calculations.

### 3.1 Creating Scalars

The `Scalar` constructor accepts multiple argument patterns:

```python
Scalar("100 J")           # String expression: value + unit
Scalar(42)                # Numeric only: dimensionless 
Scalar(100, "m")          # Value and String expression
Scalar(2, "50 J")         # Value multiplied by expression (→ 100 J)
Scalar(value=2.5, expression="W")  # Named parameters
```

- **String argument**: Parsed as complete expression (`"100 J"` → 100 Joules)
- **Numeric argument**: Creates dimensionless scalar (`42` → dimensionless 42)  
- **Two arguments**: Value multiplied by scalar from expression (`2, "50 J"` → 100 J)
- **Named parameters**: Explicit control over value and unit expression

In [80]:
from rmnpy.sitypes import Scalar

# Basic physics calculations demonstrating automatic unit tracking
distance = Scalar(100, "m")
time = Scalar(2, "s")

# Velocity = distance / time
speed = distance / time
print(f"Speed: {speed}")
print(f"Speed unit: {speed.unit}")
print(f"Speed dimensionality: {speed.dimensionality}")

# Force = mass * acceleration
mass = Scalar(5, "kg")
acceleration = Scalar(9.8, "m/s^2")
force = mass * acceleration

print(f"\nForce: {force}")
print(f"Force unit: {force.unit}")
print(f"Force dimensionality: {force.dimensionality}")

# Kinetic energy = 0.5 * mass * velocity^2
kinetic_energy = 0.5 * mass * speed**2
print(f"\nKinetic energy: {kinetic_energy}")
print(f"Energy unit: {kinetic_energy.unit}")
print(f"Energy dimensionality: {kinetic_energy.dimensionality}")

# Work = force * distance
work = force * distance
print(f"\nWork done: {work}")
print(f"Work unit: {work.unit}")
print(f"Work dimensionality: {work.dimensionality}")

# Verify energy and work have same dimensionality
print(f"\nEnergy and work have same dimensionality: {kinetic_energy.dimensionality == work.dimensionality}")

Speed: 50 m/s
Speed unit: m/s
Speed dimensionality: L/T

Force: 49 N
Force unit: N
Force dimensionality: M•L/T^2

Kinetic energy: 6250 J
Energy unit: J
Energy dimensionality: M•L^2/T^2

Work done: 4900 J
Work unit: J
Work dimensionality: M•L^2/T^2

Energy and work have same dimensionality: True
